keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,920 @@
+* Be able to hot-plug and test new resource managers.  
+* Be able to remove some of the complexity with current setup and, over time, reduce
+  the amount of code contained within Kubelet.  Instead, build a library with specific
+  needs.
+* Have a faster path to desired changes, without potentially impacting the core of 
+  Kubernetes with every policy change. <br>
+  Note that current solutions have been cropping up to allow for resource management
+  outside the core Kubelet. These solutions require turning off the Kubelet functionality 
+  and overriding current Kubelet allocation, such as [CPU Pooler](https://github.com/nokia/CPU-Pooler). 
+  We should provide a path otherwise. Many of these solutions are custom to particular companies. 
+  Many are not open source. The community should give a process to allow this functionality in core
+  kubelet at a granular level, same as we have in many systems, such as HPC or telco.
+* Leverage additional parts from the pod spec information which could help make allocation decisions
+without the need to query the k8s controlplane.
+* Be able to do research, with minimum toil, on new policies and resource management strategies
+
+This design will also use gRPC, which is used for many other pluggable components within Kubernetes.  
+The hope is that it will also, as a side effect, allow a path to simplify the Kubelet as it is 
+today and move the complexity into external plugins.
+
+### Goals
+
+#### Alpha Goals
+
+1.  Provide Initial infrastructure to support CCI Driver Plugins and k8s Deployments requiring CCI Drivers after scheduling
+1.  Provide a feature gate to enable CCI Manager. The feature will require CPU Manager policy set to None.
+1.  Provide a CCI test driver which can demonstrate CCI Driver implementation requirementes and several resource management use-cases for cpu and memory.
+1.  Provide documentation of the CCI Manager and the provided test driver plus illustration of the currently covered use-cases, including a sample driver.
+1.  Provide end-to-end tests using the sample driver.
+1.  Support seamless k8s start.
+1.  CCI Driver results (cpusets and memory affinity) are passed back to kubelet and allocated.
+1.  Handle resource management state for cpu and memory through CCI. 
+1.  Support proper fail mechanisms of Pods if CCI driver is not available - Error message + retry 
+1.  Continue cluster operation if CCI driver is not available
+1.  Podspecs:  Will be able to support current pod specs.  While there may be additional 
+extensibility in the future, the current pod specs will still work.
+1.  Guarantee that cpu and memory resources managed by CCI are visible for the scheduler to be able to 
+    correctly assign nodes.  This can be limited as far as simple CPU and memory counts for this version.
+
+#### Beta & Post-Beta Goals
+
+1.  Support standard pod deployments not requiring CCI Drivers through CPU & Memory Manager: all QoS types
```

Comment:
Have I understood correctly that this KEP doesn't propose replacing existing CPU & Memory manager, even in future? It basically proposes to add yet another manager and keep it in the Kubelet, correct?

Reply:
We would like to consider in post-alpha phase how we can handle deployments of majority of todays use-cases which do not require special handling of resources through a fast path not going through drivers and still have CCI drivers available  to users who need them and with that also come back to the refactoring topic.

Reply:
In other words, @bart0sh, the point (over time) is to start moving complex cpu/memory management out of kubelet that is not used in the majority of use cases, but allow flexibility for custom solutions for cases where performance and/or sustainability are priorities.  We can also use this path to consolidate code internally and do some cleanup.

Reply:
@catblade If this is the case, I'd suggest to clearly state it in the KEP.

Reply:
will be fixed in text